File: backend/src/NLC.Infrastructure/Services/PlanningService.cs
using Microsoft.EntityFrameworkCore;
using NLC.Application.DTOs.Planning;
using NLC.Core.Entities;
using NLC.Core.Enums;
using NLC.Infrastructure.Data;

namespace NLC.Infrastructure.Services;

public class PlanningService(NlcDbContext db)
{
    public async Task<List<PlanningSlotDto>> GetSlotsAsync(
        Guid? warehouseId, DateOnly? from, DateOnly? to, Guid[] userWarehouseIds, bool isAdmin)
    {
        var q = db.PlanningSlots.Include(p => p.Warehouse).AsQueryable();

        if (!isAdmin && userWarehouseIds.Length > 0)
            q = q.Where(p => userWarehouseIds.Contains(p.WarehouseId));

        if (warehouseId.HasValue) q = q.Where(p => p.WarehouseId == warehouseId);
        if (from.HasValue)        q = q.Where(p => p.SlotDate >= from);
        if (to.HasValue)          q = q.Where(p => p.SlotDate <= to);

        return await q
            .OrderBy(p => p.SlotDate).ThenBy(p => p.SlotTime)
            .Select(p => new PlanningSlotDto(
                p.Id, p.WarehouseId, p.Warehouse.Name,
                p.SlotDate, p.SlotTime, p.JobType,
                p.ShipmentDescription, p.ContainerNumber,
                p.AsnNumber, p.CustomerName, p.DriverName,
                p.Status, p.ErpReference))
            .ToListAsync();
    }

    public async Task<PlanningSlotDto> CreateAsync(CreatePlanningSlotRequest req)
    {
        var slot = new PlanningSlot
        {
            WarehouseId         = req.WarehouseId,
            SlotDate            = req.SlotDate,
            SlotTime            = req.SlotTime,
            JobType             = req.JobType,
            ShipmentDescription = req.ShipmentDescription,
            ContainerNumber     = req.ContainerNumber,
            AsnNumber           = req.AsnNumber,
            CustomerName        = req.CustomerName,
            DriverName          = req.DriverName,
            Status              = PlanningSlotStatus.PLANNED,
            ErpReference        = req.ErpReference,
        };
        db.PlanningSlots.Add(slot);
        await db.SaveChangesAsync();

        var wh = await db.Warehouses.FindAsync(slot.WarehouseId);
        return new PlanningSlotDto(
            slot.Id, slot.WarehouseId, wh!.Name,
            slot.SlotDate, slot.SlotTime, slot.JobType,
            slot.ShipmentDescription, slot.ContainerNumber,
            slot.AsnNumber, slot.CustomerName, slot.DriverName,
            slot.Status, slot.ErpReference);
    }

    public async Task<(bool ok, string error)> UpdateAsync(Guid id, UpdatePlanningSlotRequest req)
    {
        var slot = await db.PlanningSlots.FindAsync(id);
        if (slot is null) return (false, "Planning slot not found");

        slot.SlotDate            = req.SlotDate;
        slot.SlotTime            = req.SlotTime;
        slot.ShipmentDescription = req.ShipmentDescription;
        slot.ContainerNumber     = req.ContainerNumber;
        slot.AsnNumber           = req.AsnNumber;
        slot.CustomerName        = req.CustomerName;
        slot.DriverName          = req.DriverName;
        slot.Status              = req.Status;
        slot.ErpReference        = req.ErpReference;

        await db.SaveChangesAsync();
        return (true, string.Empty);
    }

    public async Task<(bool ok, string error)> DeleteAsync(Guid id)
    {
        var slot = await db.PlanningSlots.FindAsync(id);
        if (slot is null) return (false, "Planning slot not found");
        db.PlanningSlots.Remove(slot);
        await db.SaveChangesAsync();
        return (true, string.Empty);
    }
}

File: backend/src/main/java/com/nlc/service/PlanningService.java
package com.nlc.service;

import com.nlc.domain.entity.PlanningSlot;
import com.nlc.domain.enums.Enums.PlanningSlotStatus;
import com.nlc.repository.PlanningSlotRepository;
import com.nlc.repository.WarehouseRepository;
import com.nlc.service.JobCardService.ServiceResult;
import com.nlc.web.dto.PlanningDtos.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class PlanningService {

    private final PlanningSlotRepository slotRepo;
    private final WarehouseRepository warehouseRepo;

    public List<PlanningSlotDto> getSlots(UUID warehouseId, LocalDate from, LocalDate to,
                                           String[] userWarehouseIds, boolean isAdmin) {
        LocalDate rangeFrom = from != null ? from : LocalDate.now().minusMonths(1);
        LocalDate rangeTo   = to   != null ? to   : LocalDate.now().plusMonths(2);

        List<PlanningSlot> slots;
        if (warehouseId != null) {
            slots = slotRepo.findByWarehouseIdAndSlotDateBetweenOrderBySlotDateAscSlotTimeAsc(
                warehouseId, rangeFrom, rangeTo);
        } else {
            slots = slotRepo.findBySlotDateBetweenOrderBySlotDateAscSlotTimeAsc(rangeFrom, rangeTo);
        }

        return slots.stream()
            .filter(s -> {
                if (!isAdmin && userWarehouseIds != null && userWarehouseIds.length > 0)
                    return Arrays.asList(userWarehouseIds).contains(s.getWarehouseId().toString());
                return true;
            })
            .map(s -> new PlanningSlotDto(
                s.getId(), s.getWarehouseId(),
                s.getWarehouse() != null ? s.getWarehouse().getName() : "",
                s.getSlotDate(), s.getSlotTime(), s.getJobType(),
                s.getShipmentDescription(), s.getContainerNumber(),
                s.getAsnNumber(), s.getCustomerName(), s.getDriverName(),
                s.getStatus().name(), s.getErpReference()
            ))
            .collect(Collectors.toList());
    }

    @Transactional
    public PlanningSlotDto create(CreatePlanningSlotRequest req) {
        PlanningSlot slot = PlanningSlot.builder()
            .warehouseId(req.warehouseId())
            .slotDate(req.slotDate())
            .slotTime(req.slotTime())
            .jobType(req.jobType())
            .shipmentDescription(req.shipmentDescription())
            .containerNumber(req.containerNumber())
            .asnNumber(req.asnNumber())
            .customerName(req.customerName())
            .driverName(req.driverName())
            .status(PlanningSlotStatus.PLANNED)
            .erpReference(req.erpReference())
            .build();

        slotRepo.save(slot);

        var wh = warehouseRepo.findById(req.warehouseId()).orElse(null);
        return new PlanningSlotDto(
            slot.getId(), slot.getWarehouseId(),
            wh != null ? wh.getName() : "",
            slot.getSlotDate(), slot.getSlotTime(), slot.getJobType(),
            slot.getShipmentDescription(), slot.getContainerNumber(),
            slot.getAsnNumber(), slot.getCustomerName(), slot.getDriverName(),
            slot.getStatus().name(), slot.getErpReference()
        );
    }

    @Transactional
    public ServiceResult update(UUID id, UpdatePlanningSlotRequest req) {
        PlanningSlot slot = slotRepo.findById(id).orElse(null);
        if (slot == null) return ServiceResult.failure("Planning slot not found");

        slot.setSlotDate(req.slotDate());
        slot.setSlotTime(req.slotTime());
        slot.setShipmentDescription(req.shipmentDescription());
        slot.setContainerNumber(req.containerNumber());
        slot.setAsnNumber(req.asnNumber());
        slot.setCustomerName(req.customerName());
        slot.setDriverName(req.driverName());
        slot.setErpReference(req.erpReference());
        if (req.status() != null)
            slot.setStatus(PlanningSlotStatus.valueOf(req.status().toUpperCase()));

        slotRepo.save(slot);
        return ServiceResult.success();
    }

    @Transactional
    public ServiceResult delete(UUID id) {
        if (!slotRepo.existsById(id)) return ServiceResult.failure("Planning slot not found");
        slotRepo.deleteById(id);
        return ServiceResult.success();
    }
}
